libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -630,6 +630,104 @@ string YulUtilFunctions::overflowCheckedIntExpFunction(
+{
+	solAssert(!_exponentType.isSigned(), "");
+	solAssert(_baseType.isNegative() == _commonType.isSigned(), "");
+	solAssert(_commonType.numBits() == 256, "");
+
+	// Calculates the upperbound for exponentiation, that is, calculate `b`, such that
+	// _base**b <= _maxValue and _base**(b + 1) > _maxValue
+	auto findExponentUpperbound = [](bigint const _base, bigint const _maxValue) -> unsigned
+	{
+		unsigned first = 0;
+		unsigned last = 255;
+		unsigned middle;
+
+		while (first < last)
+		{
+			middle = (first + last) / 2;
+
+			if (
+				// The condition on msb is a shortcut that avoids computing large powers in
+				// arbitrary precision.
+				boost::multiprecision::msb(_base) * middle <= boost::multiprecision::msb(_maxValue) &&
+				boost::multiprecision::pow(_base, middle) <= _maxValue
+			)
+			{
+				if (boost::multiprecision::pow(_base, middle + 1) > _maxValue)
+					return middle;
+				else
+					first = middle + 1;
+			}
+			else
+				last = middle;
+		}
+
+		return last;
+	};
+
+
+	u256 baseValue = _baseType.literalValue(nullptr);
+	unsigned exponentUpperbound;
+
+	if (_baseType.isNegative())
+	{
+		s256 signedValue = u2s(baseValue);
+		// Only checks for underflow. The only case where this can be a problem is when, for a
+		// negative base, say `b`, and an even exponent, say `e`, `b**e = 2**255` (which is an
+		// overflow.) But this never happens because, `255 = 3*5*17`, and therefore there is no even
+		// number `e` such that `b**e = 2**255`.
+		exponentUpperbound = findExponentUpperbound(abs(signedValue), abs(_commonType.minValue()));
+
+		// That is, the power `baseValue**exponentUpperbound` is positive
+		if (exponentUpperbound % 2 == 0)
+			solAssert(
+				boost::multiprecision::pow(bigint(signedValue), exponentUpperbound) <= _commonType.maxValue() &&
+				// underflow
+				boost::multiprecision::pow(bigint(signedValue), exponentUpperbound + 1) < _commonType.minValue(),
+				"Incorrect exponent upper bound calculated."
+			);
+		else
+			solAssert(
+				boost::multiprecision::pow(bigint(signedValue), exponentUpperbound) >= _commonType.minValue() &&
+				boost::multiprecision::pow(bigint(baseValue), exponentUpperbound + 1) > _commonType.maxValue(),
+				"Incorrect exponent upper bound calculated."
+			);
+	}
+	else
+	{
+		exponentUpperbound = findExponentUpperbound(baseValue, _commonType.maxValue());
+
+		solAssert(
+			boost::multiprecision::pow(bigint(baseValue), exponentUpperbound) <= _commonType.maxValue() &&
+			boost::multiprecision::pow(bigint(baseValue), exponentUpperbound + 1) > _commonType.maxValue(),
+			"Incorrect exponent upper bound calculated."
+		);
+	}
+
+	string functionName = "checked_exp_" + _baseType.richIdentifier() + "_" + _exponentType.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(exponent) -> power {
+				exponent := <exponentCleanupFunction>(exponent)
+				if gt(exponent, <exponentUpperbound>) { revert(0, 0) }
```

Comment:
please adjust to call the panic function

Reply:
I'll make a PR that targets breaking where it calls panic function, after this gets merged.

Reply:
Didn't realize there was `panic_error` in develop already. I changed revert to panic.